### PR TITLE
Use tempfile for creating a file in home directory for a IO test

### LINKED ISSUE #40091: BUG: IO unit test creates file at home directory
- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the latest version of pandas.
- [X] (optional) I have confirmed this bug exists on the master branch of pandas.
---

The unit test at `pandas/tests/io/test_common.py` produces a file using a hardcoded name as shown below:
```python
filename = path_type("~/sometest")
with icom.get_handle(filename, "w") as handles:
  assert os.path.isabs(handles.handle.name)
  assert os.path.expanduser(filename) == handles.handle.name
```

#### Problem description
A developer may encounter an empty file "sometest" appearing in their home directory (after running `pytest`). If the developer is unaware of that, then it may raise some wrong security concerns on their side. Also, in the extreme scenario that the developer has a file at the exact same path, then its content may be replaced.

#### Expected Output
I would expect the file produced by the unit test to be temporary and get deleted afterwards. Also, that the filename is generated rather than hardcoded.

### PR DESCRIPTION
**Pull request by chrispe**
- [X] closes #40091
- [X] tests added / passed
- [X] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [X] whatsnew entry

### CODE REVIEW

File: pandas/tests/io/test_common.py [hunk 1]
```diff
@@ -119,10 +121,11 @@ def test_infer_compression_from_path(self, extension, expected, path_type):
     @pytest.mark.parametrize("path_type", [str, CustomFSPath, Path])
     def test_get_handle_with_path(self, path_type):
         # ignore LocalPath: it creates strange paths: /absolute/~/sometest
-        filename = path_type("~/sometest")
-        with icom.get_handle(filename, "w") as handles:
-            assert os.path.isabs(handles.handle.name)
-            assert os.path.expanduser(filename) == handles.handle.name
+        with tempfile.NamedTemporaryFile(dir=os.path.expanduser("~/")) as tmp:
```
Comment (jreback) [inline pandas/tests/io/test_common.py:124]: use tm.ensure_clean

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/io/test_common.py (hunk 1 above)
Comment (chrispe) [reply pandas/tests/io/test_common.py:124]: I agree, but current implementation of `tm.ensure_clean` doesn't support custom directories. Should we then first extend that function?

Comment (jreback) [reply pandas/tests/io/test_common.py:124]: ok i guess this is fine. can you use pathlib.Path instead

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm ping on greenish (there is a failing numpy_dev build ATM)

**Comment by jreback** [comment]
thanks @chrispe very nice!

**Pull request merged**